Makefile

```diff
@@ -98,147 +106,129 @@ prepare-chrome-beta:
 
-# create build dir ready for source
-ifeq ('$(browser)','chrome-mv3')
-$(BUILD_DIR)/public/js: shared/data/bundled/smarter-encryption-rules.json
-else
-$(BUILD_DIR)/public/js:
-endif
-	mkdir -p $(BUILD_DIR)/public/js/
-	mkdir -p $(BUILD_FOLDERS)
-
-# fetch SE data for bundled SE rules
-shared/data/smarter_encryption.txt:
-	curl https://staticcdn.duckduckgo.com/https/smarter_encryption.txt.gz | gunzip -c > shared/data/smarter_encryption.txt
-
-# build SE rules for MV3
-shared/data/bundled/smarter-encryption-rules.json: shared/data/smarter_encryption.txt
-	npm run bundle-se
-
-# fetch integration test data
+# Fetch integration test data.
 integration-test/artifacts/attribution.json: node_modules/privacy-test-pages/adClickFlow/shared/testCases.json setup-artifacts-dir
 	mkdir -p integration-test/artifacts
 	cp $< $@
 
-clean:
-	rm -f shared/data/smarter_encryption.txt shared/data/bundled/smarter-encryption-rules.json integration-test/artifacts/attribution.json:
-	rm -rf $(BUILD_DIR)
-
-
-###--- Copy targets ---###
-## Targets to copy artifacts to the extension build dir
-AUTOFILL_DIR = node_modules/@duckduckgo/autofill/dist
-DASHBOARD_DIR = node_modules/@duckduckgo/privacy-dashboard/build/app
-SURROGATES_DIR = node_modules/@duckduckgo/tracker-surrogates/surrogates
-BUILD_FOLDERS = $(BUILD_DIR)/public/js/content-scripts $(BUILD_DIR)/public/css
-BROWSER_TYPE = $(browser)
-COPY_DIRS = $(BUILD_DIR)/manifest.json
-ifeq ($(browser),chrome-mv3)
-	BROWSER_TYPE := chrome
-	COPY_DIRS += $(BUILD_DIR)/managed-schema.json
-endif
 
-## Copy tasks: Copying resources that don't need and compiling
-$(BUILD_DIR)/manifest.json: browsers/$(browser)/*.json
-	mkdir -p `dirname $@`
-	cp browsers/$(browser)/*.json $(BUILD_DIR)
+###--- Mkdir targets ---#
+# Note: Intermediate directories can be omitted.
+MKDIR_TARGETS = $(BUILD_DIR)/_locales $(BUILD_DIR)/data $(BUILD_DIR)/html $(BUILD_DIR)/img \
+                $(BUILD_DIR)/dashboard $(BUILD_DIR)/web_accessible_resources \
+                $(BUILD_DIR)/public/js/content-scripts $(BUILD_DIR)/public/css \
+                $(BUILD_DIR)/public/font
 
-build/chrome-mv3/$(type)/managed-schema.json: browsers/chrome/managed-schema.json
-	mkdir -p `dirname $@`
-	cp $< $@
+$(MKDIR_TARGETS):
+	mkdir -p $@
 
-$(BUILD_DIR)/_locales: browsers/chrome/_locales
-	mkdir -p `dirname $@`
-	cp -r $< $@
 
-$(BUILD_DIR)/data: $(ITEMS)
-	mkdir -p `dirname $@`
-	cp -r $(ITEMS) $(BUILD_DIR)
+###--- Copy targets ---###
+# The empty $(LAST_COPY) file is used to keep track of file copying, since translating the necessary
+# copying to proper Makefile targets is problematic.
+# See https://www.gnu.org/software/make/manual/html_node/Empty-Targets.html
+LAST_COPY = build/.last-copy-$(browser)-$(type)
 
-$(BUILD_DIR)/dashboard: $(DASHBOARD_DIR)/
-	cp -r $< $@
+RSYNC = rsync -ra --delete --exclude="*~"
 
-$(BUILD_DIR)/web_accessible_resources: $(SURROGATES_DIR)/
-	cp -r $< $@
+$(LAST_COPY): $(WATCHED_FILES) | $(MKDIR_TARGETS)
```

Comment:
The old file copying approach didn't really work. Updated files often weren't updated for example. I spent some time trying different approaches, but I found that having "proper" Makefile targets for all the copied files was a pain in practice.

Far easier to use rsync to copy all the files, then take a note (via touch) of the last copy time. That way, files are only re-copied if the last copy time is older than some of the files. Rsync takes care of updated the right files, and we do anything fancy.

Reply:
I think `-a` for rsync handles this all for you, and only copies changed/new files over. I could be wrong though.

Reply:
Yea, and we also needed a way to know if copied files needed to be copied again. Like you say, rsync knows to only copy changed files, but we didn't want the watcher to constantly call rsync on every loop.